src/tree/gpu_hist/row_partitioner.cuh

```diff
@@ -83,73 +262,61 @@ class RowPartitioner {
-  std::vector<bst_node_t> GetPositionHost();
+  template <typename UpdatePositionOpT, typename OpDataT>
+  void UpdatePositionBatch(const std::vector<bst_node_t>& nidx,
+                           const std::vector<bst_node_t>& left_nidx,
+                           const std::vector<bst_node_t>& right_nidx,
+                           const std::vector<OpDataT>& op_data, UpdatePositionOpT op) {
+    if (nidx.empty()) return;
+    CHECK_EQ(nidx.size(), left_nidx.size());
+    CHECK_EQ(nidx.size(), right_nidx.size());
+    CHECK_EQ(nidx.size(), op_data.size());
 
-  /**
-   * \brief Updates the tree position for set of training instances being split
-   * into left and right child nodes. Accepts a user-defined lambda specifying
-   * which branch each training instance should go down.
-   *
-   * \tparam  UpdatePositionOpT
-   * \param nidx        The index of the node being split.
-   * \param left_nidx   The left child index.
-   * \param right_nidx  The right child index.
-   * \param op          Device lambda. Should provide the row index as an
-   * argument and return the new position for this training instance.
-   */
-  template <typename UpdatePositionOpT>
-  void UpdatePosition(bst_node_t nidx, bst_node_t left_nidx,
-                      bst_node_t right_nidx, UpdatePositionOpT op) {
-    Segment segment = ridx_segments_.at(nidx);  // rows belongs to node nidx
-    auto d_ridx = ridx_.CurrentSpan();
-    auto d_position = position_.CurrentSpan();
-    if (left_counts_.size() <= static_cast<size_t>(nidx)) {
-      left_counts_.resize((nidx * 2) + 1);
-      thrust::fill(left_counts_.begin(), left_counts_.end(), 0);
+    auto h_batch_info = pinned2_.GetSpan<PerNodeData<OpDataT>>(nidx.size());
+
+    std::size_t total_rows = 0;
+    for (int i = 0; i < nidx.size(); i++) {
+      h_batch_info[i] = {ridx_segments_.at(nidx.at(i)).segment, op_data.at(i)};
+      total_rows += ridx_segments_.at(nidx.at(i)).segment.Size();
+    }
+    static_assert(sizeof(PerNodeData<OpDataT>) * kMaxUpdatePositionBatchSize <=
+                  sizeof(constant_memory));
+    dh::safe_cuda(cudaMemcpyToSymbolAsync(constant_memory, h_batch_info.data(),
+                                          h_batch_info.size() * sizeof(PerNodeData<OpDataT>), 0,
+                                          cudaMemcpyDefault, stream_));
+
+    // Temporary arrays
+    auto h_counts = pinned_.GetSpan<PartitionCountsT>(nidx.size(), PartitionCountsT{});
+    dh::TemporaryArray<PartitionCountsT> d_counts(nidx.size(), PartitionCountsT{});
+
+    // Partition the rows according to the operator
+    SortPositionBatch<RowIndexT, UpdatePositionOpT, OpDataT>(
+        dh::ToSpan(ridx_), dh::ToSpan(ridx_tmp_), dh::ToSpan(d_counts), total_rows, op, stream_);
+    dh::safe_cuda(cudaMemcpyAsync(h_counts.data(), d_counts.data().get(),
+                                  sizeof(decltype(d_counts)::value_type) * d_counts.size(),
+                                  cudaMemcpyDefault, stream_));
```

Comment:
Wouldn't it make the copy a synchronizing operation if it's immediately followed by a stream sync? (just curious about the result, I don't know how to remove it)

Reply:
Yes unfortunately we need the result. One of my next ideas is to process the result in a cuda stream callback thread, allowing the main thread to continue.

Reply:
That experiment didn't provide much benefit. The next step will be to pass this information directly on the device to the histogram kernels, deferring the synchronisation step.